course3/homework-list/list.h

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <memory>
+
+
+namespace Karun
+{
+
+template<typename T>
+class List
+{
+public:
+	class Node
+	{
+	friend class List;
+	
+	public:
+		Node(const T& data, std::unique_ptr<Node>& next)
+		: 	mData(data)
+		{
+			mNext = std::move(next);
+		}
+		
+	private:
+		T mData;
+		std::unique_ptr<Node> mNext;
+	};
+	
+	class iterator
+	{
+	friend class List;
+		
+	public:
+		iterator(Node* node) : mNode(node) {}
+		
+		bool operator==(const iterator& it) { return mNode == it.mNode; }
+		bool operator!=(const iterator& it) { return !operator==(it); }
+		iterator& operator++() { mNode = mNode->mNext.get(); return *this; }
+		T& operator*() { return mNode->mData; }
+		
+	private:
+		Node* mNode = nullptr;
+	};
+	
+	class const_iterator
+	{
+	public:
+		const_iterator(const iterator& it) : mNode(it.mNode) {}
+		const_iterator(Node* node) : mNode(node) {}
+		
+		bool operator==(const const_iterator& it) { return mNode == it.mNode; }
+		bool operator!=(const const_iterator& it) { return !operator==(it); }
+		const_iterator& operator++() { mNode = mNode->mNext.get(); return *this; }
+		const T& operator*() { return mNode->mData; }
+		
+	private:
+		Node* mNode = nullptr;
+	};
+	
+	List() = default;
+	List(const std::initializer_list<T>& list) { for (const T& t : list) push_front(t); }
+	
+	iterator begin() { return mBegin.get(); }
+	const_iterator begin() const { return mBegin.get(); }
+	iterator end() { return nullptr; }
+	const_iterator end() const { return nullptr; }
+	
+	bool empty() const { return mSize == 0; }
+	std::size_t size() const { return mSize; }
+	
+	T& front() { return mBegin->mData; }
+	const T& front() const { return mBegin->mData; }
+	
+	void push_front(const T& data)
```

Comment:
Will do. I guess it'll break since Node will try and copy data (which is a unique_ptr) in its constructor? Perhaps I should store references instead.